File: 314. Binary Tree Vertical Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def verticalOrder(self, root: TreeNode) -> List[List[int]]:
        '''
        BFS travesal to find out which row it belong to
        Time O(n) since we traversal each node two times
        Space O(n) a queue and a dictionary for nodes
        '''
        dic = collections.defaultdict(list)
        mincol = 2**31-1
        maxcol = -2**31
        
        if not root:
            return []
        queue = [(root, 0)]
        while queue:
            node, col = queue.pop(0)
            dic[col].append(node)
            if node.left:
                queue.append((node.left, col-1))
            if node.right:
                queue.append((node.right, col+1))
            mincol = min(col, mincol)
            maxcol = max(col, maxcol)
        res = []
        for i in range(mincol, maxcol+1):
            res.append([node.val for node in dic[i]])
        return res
            

File: 314. Binary Tree Vertical Order Traversal.java
import javafx.util.Pair;
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        if (root==null) {
            return new LinkedList<>();
        }
        LinkedList<List<Integer>> ans = new LinkedList<>();
        ans.add(new ArrayList<>());
        int minIndex = 0;
        int maxIndex = 0;
        Queue<Pair<TreeNode, Integer>> q = new ArrayDeque<>();
        q.offer(new Pair(root, 0));
        while (!q.isEmpty()) {
            Pair<TreeNode, Integer> p = q.poll();
            TreeNode curr = p.getKey();
            int pos = p.getValue();
            if (pos < minIndex) {
                minIndex = pos;
                ans.addFirst(new ArrayList<>());
            }
            if (pos > maxIndex) {
                maxIndex = pos;
                ans.addLast(new ArrayList<>());
            }
            ans.get(pos-minIndex).add(curr.val);
            if (curr.left != null) {
                q.offer(new Pair(curr.left, pos-1));
            }
            if (curr.right != null) {
                q.offer(new Pair(curr.right, pos+1));
            }
        }
        return ans;
    }
}